src/core.js

```diff
@@ -230,19 +230,23 @@ function resumeProcessing( test ) {
 	// A slight delay to allow this iteration of the event loop to finish (more assertions, etc.)
 	if ( defined.setTimeout ) {
 		setTimeout( function() {
-			var current = test || config.current;
-			if ( current && ( current.semaphore > 0 || current.resumed ) ) {
+			var current;
+
+			// Ensure we are resuming the current test
+			if ( test && test === config.current ) {
+				current = test;
```

Comment:
That was my first assumption as well. This falls apart however when you have waterfall calls that resolve close together. Consider:

```js
test('waterfall', function(assert) {
  let done1 = assert.async();
  setTimeout(() => {
    done1();
    let done2 = assert.async();
    setTimeout(done2);
  });
});
```

The resolution order can potentially go like this:
1. `done1` --> `resumeProcessing` --> sechdule `setTimeout` for 13ms
2. `setTimeout` in `test` --> `done2` --> `resumeProcessing` --> schedule `setTimeout` for 13ms
3. `setTimeout` from `done1` triggers and resumes processing
4. `resumed` is cleared as next `test` begins
5. `setTimeout` from `done2` triggers and attempts to resume processing since it doesn't know that it has already been `resumed`